pkcs11/_pkcs11.pyx

```diff
@@ -348,7 +360,7 @@ class Session(types.Session):
             Attribute.TOKEN: store,
             Attribute.PRIME_BITS: param_length,
         }
-        template_.update(template or {})
+        _finalize_template(template_, template)
```

Comment:
You conceivably could then remove all of the weird underscore confusion with something like

```python
attrs = AttributeList(merge_templates({
    default: template,
    goes: here,
}, template))
```

Which I think would make the code a lot more readable.

You could also try making this a `@classmethod` on `AttributeList` (e.g. `from_template`). Doing this means you would have to deal with Cython's binding nonsense, [see the tests for examples](https://github.com/cython/cython/blob/master/tests/run/classmethod.pyx).